### PR TITLE
Add TLS support to Exhibitor

### PR DESCRIPTION
**Pull request by rdeliallisi**
Add TLS support to Exhibitor. The support is added to both Client and Server and is configurable through environment variables. Related JIRA ticket https://jira.mesosphere.com/browse/DCOS-17894.

# Changes
* Added `HttpsConfiguration.java` which is responsible for containing both the Server and Client configuration and checking if it is valid. Unit tests in `TestHttpsConfiguration.java`.
* Added `ExhibitorEnv.java` which just contains the environment argument names that are used. TODO: Might be merged into `HttpsConfiguration.java`.
* Added `RemoteInstanceRequestHttpsClientImpl.java` which is an HTTPS Jersey REST client implementation. Unit tests in `TestRemoteInstanceRequestClient.java`
* Minimal changes to other files to add TLS support.

### CODE REVIEW

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/automanage/RemoteInstanceRequestHttpsClientImpl.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            {
+                public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException
+                {
+                }
+
+                public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException
+                {
+                }
+
+                public java.security.cert.X509Certificate[] getAcceptedIssuers()
```
Comment (mhrabovcin) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/automanage/RemoteInstanceRequestHttpsClientImpl.java]: Can you can use just `X509Certificate[]` here.

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                requireClientCertValue = httpsConf.verifyPeerCert;
+            }
+
+            httpsConf = new HttpsConfiguration(httpsConf.serverKeystorePath, httpsConf.serverKeystorePassword,
+                httpsConf.clientKeystorePath, httpsConf.clientKeystorePassword, httpsConf.truststorePath,
+                httpsConf.truststorePassword, requireClientCertValue, httpsConf.verifyPeerCert);
+            return this;
+        }
+
+        public Builder verifyPeerCert(String verifyPeerCertString)
```
Comment (mhrabovcin) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java]: Maybe we could also have a method that accepts boolean.

File: exhibitor-standalone/src/main/java/com/netflix/exhibitor/application/ExhibitorMain.java [hunk 3]
```diff
@@ -131,20 +132,42 @@ public ExhibitorMain(BackupProvider backupProvider, ConfigProvider configProvide
...
-        // the jetty.nio.SelectChannelConnector. This is a
-        // non-blocking I/O connector.
-        // See https://dcosjira.atlassian.net/browse/DCOS-558
-        SelectChannelConnector connector = new SelectChannelConnector();
-        connector.setPort(httpPort);
-        connector.setAcceptors(8);
-        connector.setMaxIdleTime(5000);
-        connector.setAcceptQueueSize(32);
-        server.setConnectors(new Connector[]{connector});
+        if(httpsConf.getServerKeystorePath() != null)
```
Comment (mhrabovcin) [inline exhibitor-standalone/src/main/java/com/netflix/exhibitor/application/ExhibitorMain.java]: Is no space between `if` and parenthesis code style used in exhibitor codebase?

File: exhibitor-core/src/test/java/com/netflix/exhibitor/core/automanage/JettyServer.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import org.mortbay.jetty.security.SslSocketConnector;
+import org.mortbay.jetty.servlet.Context;
+import org.mortbay.jetty.servlet.ServletHolder;
+
+import com.netflix.exhibitor.core.ExhibitorEnv;
+import com.netflix.exhibitor.core.HttpsConfiguration;
+import com.sun.jersey.api.core.DefaultResourceConfig;
+import com.sun.jersey.spi.container.servlet.ServletContainer;
+
+public class JettyServer
```
Comment (mhrabovcin) [inline exhibitor-core/src/test/java/com/netflix/exhibitor/core/automanage/JettyServer.java]: Can we add a documentation why we're adding a new server? Could this be possibly moved to tests directory only?

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/automanage/RemoteInstanceRequestHttpsClientImpl.java (hunk 1 above)
Comment (rdeliallisi) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/automanage/RemoteInstanceRequestHttpsClientImpl.java]: Yes, just did that.

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java (hunk 2 above)
Comment (rdeliallisi) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java]: Done.

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/Exhibitor.java [hunk 5]
```diff
@@ -135,7 +136,14 @@ public Exhibitor(ConfigProvider configProvider, Collection<? extends UITab> addi
         processMonitor = new ProcessMonitor(this);
         autoInstanceManagement = new RepeatingActivityImpl(log, activityQueue, QueueGroups.MAIN, new AutomaticInstanceManagement(this), getAutoInstanceManagementPeriod());
 
-        remoteInstanceRequestClient = new RemoteInstanceRequestClientImpl(arguments.remoteConnectionConfiguration);
+        if(arguments.httpsConfiguration.getServerKeystorePath() != null)
```
Comment (mhrabovcin) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/Exhibitor.java]: Should we decide here based on the **server** keystore configuration? What if we want TLS connection with peer verification but our server don't expose TLS server?

File: exhibitor-standalone/src/main/java/com/netflix/exhibitor/application/ExhibitorMain.java (hunk 3 above)
Comment (rdeliallisi) [inline exhibitor-standalone/src/main/java/com/netflix/exhibitor/application/ExhibitorMain.java]: I have to fix the code style to match that of the rest of the code base.

File: exhibitor-core/src/test/java/com/netflix/exhibitor/core/automanage/JettyServer.java (hunk 4 above)
Comment (rdeliallisi) [inline exhibitor-core/src/test/java/com/netflix/exhibitor/core/automanage/JettyServer.java]: The reason for this test server is that the original one is embedded inside `ExhibitorMain.java` and it's application configuration is dependent by `Exhibitor` itself. If this makes sense, I will add this in the class documentation.

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/Exhibitor.java (hunk 5 above)
Comment (rdeliallisi) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/Exhibitor.java]: When I discussed the overall design with JP, we agreed that setting `serverKeystorePath` enables TLS for both server and client. This is the current project document: https://docs.google.com/document/d/1JKkDWPzu_lPiF7hAMtVNptEprunClTBwzw2S-9WOVZI

**Comment by rdeliallisi** [comment]
@mhrabovcin Any more suggestions?

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            if ( verifyPeerCertString != null )
+            {
+                verifyPeerCertValue = Boolean.parseBoolean(verifyPeerCertString);
+            }
+            else
+            {
+                verifyPeerCertValue = httpsConf.verifyPeerCert;
+            }
+
+            httpsConf = new HttpsConfiguration(httpsConf.serverKeystorePath, httpsConf.serverKeystorePassword,
```
Comment (mhrabovcin) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java]: Nitpick: you could call here `return verifyPeerCert(verifyPeerCertValue)`.

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/automanage/RemoteInstanceRequestHttpsClientImpl.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2012 Netflix, Inc.
```
Comment (mhrabovcin) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/automanage/RemoteInstanceRequestHttpsClientImpl.java]: Can we use this copyright?

File: exhibitor-standalone/src/main/java/com/netflix/exhibitor/standalone/ExhibitorCreator.java [hunk 8]
```diff
@@ -230,7 +233,19 @@ else if ( "true".equalsIgnoreCase(commandLine.getOptionValue(FILESYSTEMBACKUP))
...
+        HttpsConfiguration httpsConfiguration = HttpsConfiguration.builder()
+            .serverKeystorePath(env.get(ExhibitorEnv.SERVER_KEYSTORE_PATH))
+            .serverKeystorePassword(env.get(ExhibitorEnv.SERVER_KEYSTORE_PASSWORD))
+            .clientKeystorePath(env.get(ExhibitorEnv.CLIENT_KEYSTORE_PATH))
+            .clientKeystorePassword(env.get(ExhibitorEnv.CLIENT_KEYSTORE_PASSWORD))
+            .truststorePath(env.get(ExhibitorEnv.TRUSTSTORE_PATH))
+            .truststorePassword(env.get(ExhibitorEnv.TRUSTSTORE_PASSWORD))
+            .requireClientCert(env.get(ExhibitorEnv.REQUIRE_CLIENT_CERT))
+            .verifyPeerCert(env.get(ExhibitorEnv.VERIFY_PEER_CERT))
+            .build();
```
Comment (mhrabovcin) [inline exhibitor-standalone/src/main/java/com/netflix/exhibitor/standalone/ExhibitorCreator.java]: What will happen here if none of the environment variables would be provided?

Comment (rdeliallisi) [inline exhibitor-standalone/src/main/java/com/netflix/exhibitor/standalone/ExhibitorCreator.java]: The `build` method checks if the provided configuration is valid. If nothing is passed at all the class will be build but `EXHIBITOR_TLS_SERVER_KEYSTORE_PATH' will be `null` and TLS will not be used. There is also a unit test for this case.

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/automanage/RemoteInstanceRequestHttpsClientImpl.java (hunk 7 above)
Comment (rdeliallisi) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/automanage/RemoteInstanceRequestHttpsClientImpl.java]: Has been removed now.

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java (hunk 6 above)
Comment (rdeliallisi) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java]: Done.

**Comment by rdeliallisi** [comment]
@gpaul Can you please have a look?

File: exhibitor-standalone/src/main/java/com/netflix/exhibitor/application/ExhibitorMain.java [hunk 9]
```diff
@@ -131,20 +131,42 @@ public ExhibitorMain(BackupProvider backupProvider, ConfigProvider configProvide
...
-        // See https://dcosjira.atlassian.net/browse/DCOS-558
-        SelectChannelConnector connector = new SelectChannelConnector();
-        connector.setPort(httpPort);
-        connector.setAcceptors(8);
-        connector.setMaxIdleTime(5000);
-        connector.setAcceptQueueSize(32);
-        server.setConnectors(new Connector[]{connector});
+        if ( httpsConf.getServerKeystorePath() != null )
+        {
+            SslSocketConnector connector = new SslSocketConnector();
```
Comment (gpaul) [inline exhibitor-standalone/src/main/java/com/netflix/exhibitor/application/ExhibitorMain.java]: If I remember correctly, the SslSocketConnector suffers from the same issues described in the comment below relating to the SocketConnector (ie., it has a blockling-IO implementation). Please use `SslSelectChannelConnector` and configure it with the same settings used below:
```
+            connector.setAcceptors(8);
 +            connector.setMaxIdleTime(5000);
 +            connector.setAcceptQueueSize(32);
```

**Comment by gpaul** [comment]
We'll have to keep in mind that during upgrades the upgraded exhibitors will be able to communicate exclusively among one another. 

I gave this a minute's thought and I don't think this will impact ZK: exhibitor is restarted, true, but the list of ensemble members doesn't change, nor do any ZK command line parameters, so ZK should simply keep running. This is something we should test using test-e2e, based on the test_restart.py test: https://github.com/mesosphere/dcos-enterprise/blob/f988c36f5c4db4b7fc85f961c84bc5718d312b48/test-e2e/test_restart.py.

**Comment by rdeliallisi** [comment]
@gpaul The required changes to the `SSL` connector have been made. The changes have also been made in the `Jetty` test server.

All unit tests pass. I ran an ensemble of 3 exhibitors with `SSL` turned on and `PEER` verification on. The results are:
* Exhibitors manage to talk to each other via `SSL` and ZK communication still works.
* `BASIC` authentication still works when using `SSL`.
* All `non-SSL` functionality is preserved.

File: exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+                httpsConf.clientKeystorePath, httpsConf.clientKeystorePassword, httpsConf.truststorePath,
+                httpsConf.truststorePassword, httpsConf.requireClientCert, verifyPeerCert);
+            return this;
+        }
+
+        public HttpsConfiguration build()
+        {
+            Preconditions.checkArgument(!(httpsConf.serverKeystorePath == null ^ httpsConf.serverKeystorePassword == null),
+                "Server keystore path/password must be both/neither specified.");
```
Comment (jgehrcke) [inline exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java]: @rdeliallisi and @gpaul can you work together to make these error messages be super helpful and correct English at the same time? :)

Comment (gpaul) [reply exhibitor-core/src/main/java/com/netflix/exhibitor/core/HttpsConfiguration.java]: The simplest error message would be:
```
Both or neither of EXHIBITOR_TLS_SERVER_KEYSTORE_PATH and EXHIBITOR_TLS_SERVER_KEYSTORE_PASSWORD must be specified.
```
I'd be happy to accept that change.

The same goes for the truststore check.

**Comment by rdeliallisi** [comment]
@jp All relevant error messages have been changed in accordance with @gpaul's suggestion.

**Comment by gpaul** [comment]
Would you mind adding a link to a jira ticket in the PR description?

**Comment by lingmann** [comment]
@rdeliallisi JP's github handle is @jgehrcke ... FYI :)

**Pull request merged**